### PR TITLE
Options & logger for PriDec solver

### PR DESCRIPTION
**Pull request by cnpetra**
- The PriDec solver is equipped with an options object that can load options from a file (the default file is "hiop_pridec.options) (addresses issue #279)
- Options for master and worker NLP solvers within PriDec solver can be specified independently; for master the default "hiop_pridec_master.options" file will be used (can be changed via hiop_pridec.options); while the workers can use the default 'hiop.options'
- `hiopOptions` class is specialized to work with NLP and PriDec solvers (minimal changes)
- most of `hiopOptions` class converted to the new coding styleguide 
- added option for turning on/off the printing of options
- updated logger to work with PriDec solver (address issue #249)

### CODE REVIEW

**Comment by nychiang** [comment]
@cnpetra  per our offline discussion, the interface for unsymmetric matrix is added.

File: src/Utils/hiopOptions.cpp [hunk 1]
```diff
@@ -849,9 +514,614 @@ void hiopOptions::OptionInt::print(FILE* f) const
...
+    //TODO: Frank check these and add others as needed in the primal decomposition algorithm
+    register_num_option("tolerance",
+                        1e-4,
+                        1e-14,
+                        1e-1,
+                        "Absolute error tolerance for the PriDec solver (default 1e-4)");
+    
+    //register_num_option("rel_tolerance", 0., 0., 0.1,
+    //                  "Error tolerance for the NLP relative to errors at the initial point. A null "
+    //                  "value disables this option (default 0.)");
```
Comment (nychiang) [inline src/Utils/hiopOptions.cpp:1076]: Should we remove these 3 lines from the source code?

**Review by nychiang — COMMENTED** [review]
LGTM. This make the code much clearer, and we can easily know where the memory is used.

File: doc/src_usermanual/techrep_main.tex [hunk 2]
```diff
@@ -1012,6 +1023,55 @@ \subsection{Problem preprocessing}
...
+
+\noindent \textbf{print\_options}: string option with ``yes`` or ``no`` values deciding whether the options should be printed on the output before solver (re)starts. Default value ``no''.
+
+
+\subsection{Options for PriDec solver}
+Here we list the options that are recognized by the \Hi's PriDec solver.
+
+\subsubsection{Termination criteria and output}
+
+\noindent \textbf{tolerance}: maximum (absolute)  TODO: Frank explanation here. Double values in $[10^{-14},0.1]$. Default value: $10^{-8}$.
```
[the remark this answers is not in the document]
Comment (jwang125) [reply doc/src_usermanual/techrep_main.tex:1036]: I can add the termination after merging.

**Comment by jwang125** [comment]
I've read through the files related to pridecomp solver. If there are bugs I think I will have a better idea when ex9 raja is implemented.

File: src/Interface/hiopInterfacePrimalDecomp.hpp [hunk 3]
```diff
@@ -60,39 +60,37 @@ class hiopInterfacePriDecProblem
    */
   hiopInterfacePriDecProblem(MPI_Comm comm_world=MPI_COMM_WORLD)
```
[the remark this answers is not in the document]
Comment (cnpetra) [reply src/Interface/hiopInterfacePrimalDecomp.hpp:61]: removed for now

File: src/Utils/hiopOptions.cpp (hunk 1 above)
Comment (cnpetra) [reply src/Utils/hiopOptions.cpp:1077]: these are instructions for Frank's iteration over the PriDec solver to support new options.

File: doc/src_usermanual/techrep_main.tex (hunk 2 above)
[the remark this answers is not in the document]
Comment (cnpetra) [reply doc/src_usermanual/techrep_main.tex:1036]: Frank or I will follow up to implement this option (and a couple others) in PriDec solver

File: doc/src_usermanual/techrep_main.tex [hunk 4]
```diff
@@ -799,24 +799,35 @@ \subsection{Compiling and linking your project with the \Hi library}
...
 \end{lstlisting}
 \end{itemize}
 
-\warningcp{Note:} Options set in \texttt{hiop.options} file at runtime overwrite options set at runtime via the above API.
+\warningcp{} Each option i. should be of one of types numeric/double, integer, and string; ii. has a value associated; iii. may have a range of values;  and, iv. has a default value. 
+
+The NLP solvers load options from the file \texttt{hiop.options}. The PriDec solver will look for and load options from up to three files:
+\begin{itemize}
+\item \texttt{hiop\_pridec.options} specifies options for the \textbf{PriDec} algorithm/solver
+\item \texttt{hiop\_pridec\_master.options} specifies options for the NLP solver used to solve the \textbf{master} problem. This master NLP solver does not necessarily have to be one of \Hi's NLP solvers. The name of this file can be controlled via the string option \texttt{options\_file\_master\_prob} of the PriDec solver.
```
[the remark this answers is not in the document]
Comment (cnpetra) [reply doc/src_usermanual/techrep_main.tex]: good catch

**Review by nychiang — APPROVED** [review]
LGTM. 👍

**Pull request merged**